File: othello.py
"""
Módulo principal del juego Othello
Implementa la lógica del juego y dos niveles de IA.
"""

def leerTablero(lineas):
    """
    Construye el tablero 8x8 a partir de las líneas del archivo.
    Lanza ValueError si alguna fila no tiene 8 columnas.
    """
    tablero = []
    for i in range(8):
        fila = list(lineas[i])
        if len(fila) != 8:
            raise ValueError(f"La fila {i+1} del tablero no tiene 8 columnas")
        tablero.append(fila)

    return tablero

def leerColorInicial(color):
    """
    Valida y devuelve el color inicial ('B' o 'N').
    Lanza ValueError si es inválido.
    """
    if color not in ("B", "N"):
        raise ValueError("El color inicial debe ser 'B' o 'N'")

    return color


def imprimirTablero(tablero):
    for i in range(8):
        for j in range(8):
            print(tablero[i][j], end=" ")

        print()


def convertirJugada(jugada):
    """
    Convierte una jugada tipo "D3" a (fila, columna).
    Devuelve None si el formato es inválido.
    """
    if len(jugada) != 2:
        return None

    col = jugada[0].upper()
    fil = jugada[1]

    # Validar que la columna esté entre A y H
    if col < 'A' or col > 'H':
        return None
    # Validar que la fila esté entre 1 y 8
    if fil < '1' or fil > '8':
        return None

    # Convertir a índices de matriz (0-7)
    columna = ord(col) - ord('A')
    fila = int(fil) - 1

    return fila, columna


def jugadaValida(tablero, fila, col, color):
    """
    Verifica si una jugada es válida según las reglas de Othello.
    Una jugada es válida si:
    1. La casilla está vacía ('X')
    2. En al menos una dirección hay fichas del oponente seguidas por una ficha propia
    """
    # La casilla debe estar vacía
    if tablero[fila][col] != 'X':
        return False

    oponente = 'B' if color == 'N' else 'N'

    # Las 8 direcciones posibles
    dirs = [(-1,-1), (-1,0), (-1,1), (0,-1), (0,1), (1,-1), (1,0), (1,1)]

    valido = False

    # Revisar cada dirección
    for df, dc in dirs:
        f = fila + df
        c = col + dc
        encontrado_oponente = False

        # Avanzar mientras encontremos fichas del oponente
        while 0 <= f < 8 and 0 <= c < 8 and tablero[f][c] == oponente:
            encontrado_oponente = True
            f += df
            c += dc

        # Si encontramos al menos una ficha del oponente y luego una ficha propia, es válida
        if encontrado_oponente and 0<= f < 8 and 0 <= c < 8 and tablero[f][c] == color:
            valido = True
            break

    return valido


def aplicarJugada(tablero, fila, col, color):
    # Colocar la ficha en la posición indicada
    tablero[fila][col] = color
    oponente = 'B' if color == 'N' else 'N'

    # Las 8 direcciones posibles
    dirs = [(-1,-1), (-1,0), (-1,1), (0,-1), (0,1), (1,-1), (1,0), (1,1)]

    # Para cada dirección, voltear las fichas correspondientes
    for df, dc in dirs:
        f = fila + df
        c = col + dc
        dar_vuelta = []

        # Recolectar fichas del oponente en esta dirección
        while 0 <= f < 8 and 0 <= c < 8 and tablero[f][c] == oponente:
            dar_vuelta.append((f,c))
            f += df
            c += dc

        # Si encontramos una ficha propia al final, voltear todas las del oponente
        if dar_vuelta and 0 <= f < 8 and 0 <= c < 8 and tablero[f][c] == color:
            for (fx, cx) in dar_vuelta:
                tablero[fx][cx] = color


def turnoJugador(tablero, color):
    """
    Gestiona el turno del jugador humano.
    Solicita jugadas hasta que una sea válida.
    """
    while True:
        jugada = input("Ingrese su jugada (ej: D3): ").strip()

        coords = convertirJugada(jugada)
        if coords is None:
            print("Formato inválido, intente nuevamente")
            continue

        fila, col = coords
        if jugadaValida(tablero, fila, col, color):
            aplicarJugada(tablero, fila, col, color)
            
            print("Tablero después de su jugada:")
            imprimirTablero(tablero)
            return

        print("Jugada inválida, intente otra")


import random


def jugadasPosibles(tablero, color):
    """
    Devuelve una lista de todas las jugadas válidas para un color.
    """
    posibles = []
    for f in range(8):
        for c in range(8):
            if jugadaValida(tablero, f, c, color):
                posibles.append((f, c))
    return posibles


def nivel0(tablero, color):
    """
    IA nivel 0: elige una jugada válida al azar.
    """
    posibles = jugadasPosibles(tablero, color)
    if not posibles:
        return
    
    # Elegir aleatoriamente entre las jugadas posibles
    jugada = random.choice(posibles)
    aplicarJugada(tablero, jugada[0], jugada[1], color)

    # Convertir coordenadas a notación algebraica para mostrar
    colLetra = chr(jugada[1] + ord('A'))
    filNum = jugada[0] + 1
    print(f"La computadora eligió: {colLetra}{filNum}")
    imprimirTablero(tablero)
        

def contarFichasVueltas(tablero, fila, col, color):
    # Si la casilla no está vacía, no se puede jugar ahí
    if tablero[fila][col] != 'X':
        return 0

    oponente = 'B' if color == 'N' else 'N'

    # Las 8 direcciones posibles
    dirs = [(-1,-1), (-1,0), (-1,1), (0,-1), (0,1), (1,-1), (1,0), (1,1)]
    total = 0

    # Contar fichas volteadas en cada dirección
    for df, dc in dirs:
        f = fila + df
        c = col + dc
        contador = 0

        # Contar fichas del oponente consecutivas en esta dirección
        while 0 <= f < 8 and 0 <= c < 8 and tablero[f][c] == oponente:
            contador += 1
            f += df
            c += dc

        # Si encontramos una ficha propia al final, estas fichas se voltearían
        if 0 <= f < 8 and 0 <= c < 8 and tablero[f][c] == color:
            total += contador

    return total


def elegirNivel1(tablero, color):
    """
    Elige la jugada que maximiza fichas volteadas.
    Devuelve None si no hay jugadas posibles.
    """
    posibles = jugadasPosibles(tablero, color)

    if not posibles:
        return None
    
    mejor_jugada = None
    max_fichas = -1

    # Evaluar cada jugada posible
    for fila, col in posibles:
        vueltas = contarFichasVueltas(tablero, fila, col, color)
        if max_fichas < vueltas:
            max_fichas = vueltas
            mejor_jugada = (fila, col)
    
    return mejor_jugada


def nivel1(tablero, color):
    """
    IA nivel 1: estrategia maximiza fichas capturadas.
    """
    mejor_jugada = elegirNivel1(tablero, color) 
    
    aplicarJugada(tablero, mejor_jugada[0], mejor_jugada[1], color)

    # Convertir coordenadas a notación algebraica para mostrar
    colLetra = chr(mejor_jugada[1] + ord('A'))
    filNum = mejor_jugada[0] + 1
    print(f"La computadora eligió: {colLetra}{filNum}")
    imprimirTablero(tablero)


def contadorFichas(tablero, color_jugador, color_pc):
    cant_fichas_j = 0 
    cant_fichas_pc = 0
    for i in range(8):
        for j in range(8):
            if tablero[i][j] == color_jugador:
                cant_fichas_j += 1
            elif tablero[i][j] == color_pc:
                cant_fichas_pc += 1

    return  cant_fichas_j, cant_fichas_pc


def mostrarResultadoFinal(tablero, color_jugador, color_pc):
    cant_j, cant_pc = contadorFichas(tablero, color_jugador, color_pc)

    print("\nFin de la partida")
    print(f"Fichas jugador ({color_jugador}): {cant_j}")
    print(f"Fichas computadora ({color_pc}): {cant_pc}")

    # Determinar y mostrar el ganador
    if cant_j > cant_pc:
        print("Ganó el jugador")
    elif cant_pc > cant_j:
        print("Ganó la computadora")
    else:
        print("Empate")


def ejecutarPartida(tablero, turno, color_jugador, color_pc, modo):
    """
    Ejecuta el bucle principal del juego alternando turnos
    """
    while True:
        puede_jugador = jugadasPosibles(tablero, color_jugador)
        puede_pc = jugadasPosibles(tablero, color_pc)

        # Si ninguno puede jugar → fin del juego
        if not puede_jugador and not puede_pc:
            break

        # Si el turno actual no puede jugar → pasar turno
        if turno == color_jugador and not puede_jugador:
            print(f"{turno} no tiene jugadas, se salta el turno")
            turno = color_pc
            continue

        if turno == color_pc and not puede_pc:
            print(f"{turno} no tiene jugadas, se salta el turno")
            turno = color_jugador
            continue

        # Ejecutar turno
        if turno == color_jugador:
            turnoJugador(tablero, turno)
        else:
            if modo == "0":
                nivel0(tablero, turno)
            else:
                nivel1(tablero, turno)

        # Cambiar turno
        turno = color_pc if turno == color_jugador else color_jugador

    mostrarResultadoFinal(tablero, color_jugador, color_pc)


def main(nombre_archivo, color_jugador, modo):
    """
    Inicializa el juego y ejecuta la partida completa.
    """
    with open(nombre_archivo, "r") as f:
        lineas = [line.strip() for line in f.readlines()]

        if len(lineas) < 9:
            raise ValueError("El archivo no contiene un tablero válido.")
    
    # Leer las primeras 8 líneas como el tablero    
    tablero = leerTablero(lineas[:8])
    # La línea 9 indica qué color comienza
    turno = leerColorInicial(lineas[8])

    print("Tablero inicial:")
    imprimirTablero(tablero)

    # Determinar el color de la computadora
    color_pc = 'B' if color_jugador == 'N' else 'N'

    ejecutarPartida(tablero, turno, color_jugador, color_pc, modo)

import sys

# Punto de entrada del programa
if __name__ == "__main__":
    # Verificar que se proporcionen los argumentos correctos
    if len(sys.argv) != 4:
        print("Uso: ./othello.py archivo color modo")
        sys.exit(1)

    # Ejecutar el juego con los parámetros proporcionados
    # sys.argv[1]: archivo con tablero inicial
    # sys.argv[2]: color del jugador ('B' o 'N')
    # sys.argv[3]: modo de dificultad de la IA ('0' o '1')
    main(sys.argv[1], sys.argv[2], sys.argv[3])

File: othello.c
/*
    othello.c - Programa principal para procesar partidas de Othello
    
    Este programa lee un archivo con información de una partida de Othello
    y procesa cada movimiento, validando su legalidad y actualizando el tablero. Al finalizar, muestra el
    estado del juego y determina si continúa o si hay un ganador.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "tablero.h"

/*
    convertirJugadas
    Convierte una jugada tipo "D3" a coordenadas (fila, columna).
    Devuelve 1 si es válida, 0 si el formato es incorrecto.
 */
int convertirJugadas(const char *jugada, int *fila, int *columna) {
    /* Verificar que la jugada tenga al menos 2 caracteres */
    if (strlen(jugada) < 2) {
        return 0;
    }

    char col = jugada[0];
    char fil = jugada[1];

    if (col >= 'a' && col <= 'h') {
        col = col - 'a' + 'A';
    }

    /* Validar que la columna esté entre A y H */
    if (col < 'A' || col > 'H') {
        return 0;
    }

    /* Validar que la fila esté entre 1 y 8 */
    if (fil < '1' || fil > '8') {
        return 0;
    }

    /* Convertir a índices de matriz (0-7) */
    *columna = col - 'A';
    *fila = fil - '1';

    return 1;
}

/*
    leerJugador
    Lee una línea "Nombre,Color" del archivo.
    Asigna memoria para el nombre y valida el color ('B' o 'N').
    Devuelve 1 si es correcto, 0 si hay error.
 */
int leerJugador(FILE *archivo, char **nombre, char *color) {
    char buffer[256];
    char nombreTmp[256];
    char colorTmp;

    /* Leer línea del archivo */
    if (!fgets(buffer, sizeof(buffer), archivo))
        return 0;

    /* Parsear formato "Nombre,Color" */
    if (sscanf(buffer, "%255[^,],%c", nombreTmp, &colorTmp) != 2)
        return 0;

    /* Validar que el color sea 'B' o 'N' */
    if (colorTmp != 'B' && colorTmp != 'N')
        return 0;

    /* Asignar memoria para el nombre */
    *nombre = malloc(strlen(nombreTmp) + 1);
    if (!*nombre)
        return 0;

    strcpy(*nombre, nombreTmp);
    *color = colorTmp;

    return 1;
}

/*
    empezo
    Lee el color del jugador que comienza ('B' o 'N').
    Devuelve el color o 0 si hay error.
 */
char empezo(FILE* archivo){
    char buffer[8];

    /* Leer línea del archivo */
    if (!fgets(buffer, sizeof(buffer), archivo))
        return 0;

    if (buffer[0] != 'B' && buffer[0] != 'N')
        return 0;

    return buffer[0];
}

/*
    verificarDatos
    Verifica nombres no vacíos, colores válidos y distintos.
    Devuelve 1 si los datos son correctos, 0 si hay error.
 */
int verificarDatos(char *nombre1, char *nombre2, char color1, char color2){
    if (!nombre1 || !nombre2) {
        printf("No se pudieron leer los nombres de los jugadores\n");
        return 0;
    }
    if(nombre1[0] == '\0'){
        printf("Nombre del primer jugador vacio\n");
        return 0;
    }
    if(nombre2[0] == '\0'){
        printf("Nombre del segundo jugador vacio\n");
        return 0;   
    }
    if(color1 != 'B' && color1 != 'N'){
        printf("Color del primer jugador invalido\n");
        return 0;
    }
    if(color2 != 'B' && color2 != 'N'){
        printf("Color del segundo jugador invalido\n");
        return 0;
    }
    if(color1 == color2){
        printf("Ambos jugadores tienen el mismo color\n");
        return 0;
    }
    return 1;
}


/*
    procesarJugadasDesdeArchivo
    Lee y procesa todas las jugadas del archivo.
    Aplica cada jugada válida al tablero.
    Devuelve 1 si todas son válidas, 0 si encuentra una inválida.
 */
int procesarJugadasDesdeArchivo(FILE *puntero_archivo, Tablero *t, char primero, char segundo, char color1, char *nombre1, char *nombre2, int *turnoFinal){
    int turno = 0;
    char buffer[4];

    char colorActual = primero;
    char colorSiguiente = segundo;

    while (1) {

        /* 1. Verificar si el jugador actual puede jugar */
        if (!existeJugada(t, colorActual)) {

            /* Si ninguno puede jugar, fin de la partida */
            if (!existeJugada(t, colorSiguiente)) {
                *turnoFinal = turno;
                return 1;
            }

            /* Pase de turno */
            char aux = colorActual;
            colorActual = colorSiguiente;
            colorSiguiente = aux;
            continue;
        }

        /* 2. Leer la jugada SOLO si puede jugar */
        if (fscanf(puntero_archivo, "%3s", buffer) == EOF) {
            *turnoFinal = turno;
            return 1;
        }

        int fila, columna;
        char *nombreActual =
            (colorActual == color1) ? nombre1 : nombre2;

        /* 3. Convertir jugada */
        if (!convertirJugadas(buffer, &fila, &columna)) {
            jugadaInvalida(t, buffer, nombreActual);
            return 0;
        }

        /* 4. Validar jugada */
        if (!esJugadaValida(t, fila, columna, colorActual)) {
            jugadaInvalida(t, buffer, nombreActual);
            return 0;
        }

        /* 5. Aplicar jugada */
        aplicarJugada(t, fila, columna, colorActual);

        turno++;
        *turnoFinal = turno;

        /* 6. Cambiar turno */
        char aux = colorActual;
        colorActual = colorSiguiente;
        colorSiguiente = aux;
    }
}


/*
    main
    Lee un archivo de partida de Othello, valida los datos,
    procesa las jugadas y muestra el estado final del juego.
 */
int main(int argc, char* argv[]){
    if(argc < 2){
        /* Verificar que se proporcionó el archivo como argumento */
        printf("Debe indicar el archivo como argumento\n");
        return 1;
    }

    FILE *puntero_archivo = fopen(argv[1], "r");
    if(puntero_archivo == NULL){
        printf("No es posible abrir este archivo");
        return 1;
    }
    
    /* Leer los nombres y colores de ambos jugadores */
    char *nombre1 = NULL;
    char *nombre2 = NULL;
    char color1 = '\0';
    char color2 = '\0';
    if (!leerJugador(puntero_archivo, &nombre1, &color1) || !leerJugador(puntero_archivo, &nombre2, &color2)) {
        printf("Error al leer los jugadores\n");
        fclose(puntero_archivo);
        return 1;
    }

    /* Verificar que los datos leídos sean válidos */
    if(!verificarDatos(nombre1, nombre2, color1, color2)){
        fclose(puntero_archivo);
        free(nombre1);
        free(nombre2);
        return 1;
    }

    /* Determinar qué jugador comenzó el juego */
    char primero = empezo(puntero_archivo);
    if (primero == '\0') {
        printf("Error al leer el color que empezó\n");
        fclose(puntero_archivo);
        free(nombre1);
        free(nombre2);
        return 1;
    }
    char segundo = (primero == color1) ? color2: color1;

    Tablero tablero;
    inicializarTablero(&tablero);

    /* Leer y procesar todas las jugadas desde la línea 4 hasta EOF */
    int turnoFinal;
    if(!procesarJugadasDesdeArchivo(puntero_archivo, &tablero, primero, segundo, color1, nombre1, nombre2, &turnoFinal)){
        fclose(puntero_archivo);
        free(nombre1);
        free(nombre2);
        return 0;
    }

    /* Determinar el estado final del juego y quién debe jugar a continuación */
    char colorActual = (turnoFinal % 2 == 0) ? primero : segundo;
    char colorSiguiente = (colorActual == primero) ? segundo : primero;
    char *nombreActual = (colorActual == color1) ? nombre1 : nombre2;
    char *nombreSiguiente = (colorSiguiente == color1) ? nombre1 : nombre2;

    estadoJuego(&tablero, colorActual, colorSiguiente, nombreActual, nombreSiguiente);

    fclose(puntero_archivo);
    free(nombre1);
    free(nombre2);
    return 0;
}